libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -0,0 +1,747 @@
+    return status;
+}
+
+static MQTTStatus_t _updateRecord( MQTTPubAckInfo_t * records,
+                                   size_t recordLen,
+                                   size_t recordIndex,
+                                   uint16_t packetId,
+                                   MQTTQoS_t qos,
+                                   MQTTPublishState_t newState,
+                                   bool shouldStore,
+                                   bool shouldDelete )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    if( shouldStore )
+    {
+        status = _addRecord( records,
+                             recordLen,
+                             packetId,
+                             qos,
+                             newState );
+    }
+    else if( recordIndex >= recordLen )
+    {
+        /* Something's wrong because the record should exist. */
+        status = MQTTBadParameter;
+    }
+    else if( shouldDelete )
+    {
+        status = _deleteRecord( records,
+                                recordLen,
+                                recordIndex );
+    }
+    else
+    {
+        /* Sanity checks. */
+        if( records[ recordIndex ].packetId != packetId )
+        {
+            status = MQTTBadParameter;
+        }
+        else if( records[ recordIndex ].qos != qos )
+        {
+            status = MQTTBadParameter;
+        }
+        else
+        {
+            records[ recordIndex ].publishState = newState;
+        }
+    }
+    
+    return status;
+}
+
+MQTTPublishState_t _MQTT_CalculateState( MQTTPublishType_t packetType,
+                                         MQTTStateOperation_t opType,
+                                         MQTTQoS_t qos,
+                                         MQTTPacketIDState_t idStatus )
+{
+    MQTTPublishState_t calculatedState = MQTTInvalidState;
+    switch( qos )
+    {
+        case MQTTQoS0:
+            if( idStatus == ID_UNKNOWN && packetType == MQTT_PUBLISH )
+            {
+                if( opType == MQTT_RESERVE )
+                {
+                    calculatedState = MQTTPublishSend;
+                }
+                else /* Sent or received. */
+                {
+                    calculatedState = MQTTPublishDone;
+                }
+            }
+            /* Anything else is an error. */
+            break;
+        case MQTTQoS1:
+            calculatedState = _calculateQoS1( packetType, opType, idStatus );
+            break;
+        case MQTTQoS2:
+            calculatedState = _calculateQoS2( packetType, opType, idStatus );
+            break;
+        default:
+            break;
+    }
+    return calculatedState;
+}
+
+MQTTStatus_t _MQTT_UpdateState( MQTTContext_t * pMqttContext,
+                                uint16_t packetId,
+                                MQTTPublishType_t packetType,
+                                MQTTStateOperation_t opType,
+                                MQTTQoS_t qos )
+{
+    MQTTStatus_t status = MQTTSuccess;
```

Comment:
Prefer to initialize this to an invalid/collision state rather than `MQTTSuccess`.